eureka2-write-server/src/main/java/com/netflix/eureka2/server/channel/RegistrationChannelImpl.java

```diff
@@ -124,55 +125,81 @@ public Source getSource() {
      */
     @Override
     public Observable<Void> register(final InstanceInfo instanceInfo) {
-        if (!moveToState(STATE.Idle, STATE.Registered) &&
-                !moveToState(STATE.Registered, STATE.Registered)) {
-            STATE currentState = state.get();
-            if (STATE.Closed == currentState) {
-                // Since channel is already closed and hence the transport, we don't need to send an error on transport.
-                return Observable.error(CHANNEL_CLOSED_EXCEPTION);
-            } else {
-                Exception exception = new IllegalStateException("Unknown state error when registering, current state: " + currentState);
-                return sendErrorOnTransport(exception).doOnTerminate(new Action0() {
-                    @Override
-                    public void call() {
-                        close();
-                    }
-                });
-            }
+        if (moveToState(STATE.Idle, STATE.Registered)) {
+            return firstRegistration(instanceInfo);
```

Comment:
We could do eager registration with initially empty subject, if id was not needed. With current API we can do this only on the first register, hence we have different code path. The return observable for update is a artificial with current API, as we do not provide per update state from the registry. I also do not think we should do that, as it would add a lot of extra cost with no benefit.